crates/matrix-sdk/src/client/mod.rs

```diff
@@ -2393,7 +2398,7 @@ impl Client {
     pub async fn sync_with_callback<C>(
         &self,
         mut sync_settings: crate::config::SyncSettings<'_>,
-        callback: impl Fn(SyncResponse) -> C,
+        callback: impl Fn(Result<SyncResponse, Error>) -> C,
```

Comment:
I am not convinced about this API change. `sync_with_callback` currently is an optimistic API and maybe we should use a different API instead for this use case rather than changing its signature. As the TODO (that isn't removed) below states...